### PR TITLE
new features

### PR DESCRIPTION
**Pull request by Roleren**

### CODE REVIEW

File: R/feature_methods2.R [hunk 1]
```diff
@@ -1,3 +1,70 @@
+#' Get inition score for a GRangesList of ORFs
```
Comment (JokingHero) [inline R/feature_methods2.R]: inition??!?!?!

File: R/feature_methods2.R [hunk 2]
```diff
@@ -1,3 +1,70 @@
...
+#'
+#' initiationScore tries to check how much each TIS region resembles, the
+#' average of the CDS TIS regions.
+#'
+#' As result there is one value per ORF:
+#' 1.000: means that ORF had no reads
```
Comment (JokingHero) [inline R/feature_methods2.R]: why the values are from 0-1.99 instead of 0-1? suspicious

File: R/feature_methods2.R [hunk 3]
```diff
@@ -1,3 +1,70 @@
...
+#' 0.000: means that ORF is identical to average of CDS
+#' 1.999: means that orf is maximum different than average of CDS
+#' @references doi: 10.1186/s12915-017-0416-0
+#' @param grl a \code{\link{GRangesList}} object with ORFs
+#' @param cds a \code{\link{GRangesList}} object with coding sequences
+#' @param tx a GrangesList of transcripts covering grl.
+#' @param footprints ribozomal footprints, given as Galignment object or
```
Comment (JokingHero) [inline R/feature_methods2.R]: ribosomal!

File: R/feature_methods2.R [hunk 4]
```diff
@@ -1,3 +1,70 @@
...
+#'                ranges = IRanges(1,85),
+#'                strand = "+")
+#' tx <- GRangesList(tx1 = tx)
+#'
+#' initiationScore(grl, cds, tx, RFP, pShifted = TRUE)
+#'
+initiationScore <- function(grl, cds, tx, footprints, pShifted = TRUE) {
+  # train average cds model
+  df <- ORFik:::riboTISCoverageProportion(cds, tx, footprints, average = TRUE,
+                                  onlyProportion = FALSE, pShifted = pShifted)
```
Comment (JokingHero) [inline R/feature_methods2.R]: formatting?

File: R/shift_footprints.R [hunk 5]
```diff
@@ -140,35 +141,43 @@ detectRibosomeShifts <- function(
...
                                 minCDS = minCDS, minThreeUTR = minThreeUTR)
   if (length(txNames) == 0) stop("No transcript has leaders and trailers of",
                                  " specified minFiveUTR, minCDS, minThreeUTR")
   cds <- GenomicFeatures::cdsBy(txdb, by = "tx", use.names = TRUE)[txNames]
 
-
+  # reduce data set to only matching seqlevels
+  cds <- keepSeqlevels(cds, unique(seqnames(footprints))[
+    unique(seqnames(footprints)) %in% unique(seqnamesPerGroup(cds, FALSE))],
+      pruning.mode = "coarse")
```
Comment (JokingHero) [inline R/shift_footprints.R]: lovely formatting, not

File: tests/testthat/test_granges_helpers.R [hunk 6]
```diff
@@ -218,4 +218,27 @@ test_that("reduceKeepAttr works as intended", {
 })
 
 
+test_that("windowPerGroup works as intended", {
 
+  gr <- GRanges(seqnames = "1",
+                       ranges = IRanges(start = c(40),
```
Comment (JokingHero) [inline tests/testthat/test_granges_helpers.R]: formatting? also, some tests for new scores would be usefull

File: R/cage_annotations.R [hunk 7]
```diff
@@ -321,9 +321,12 @@ assignTSSByCage <- function(txdb, cage, extension = 1000,
 
   cds <- cdsBy(txdb,"tx",use.names = TRUE)
   cds01 <- cds[!(names(cds) %in% names(fiveUTRs))]
+  if(length(cds01) == 0) {
```
Comment (JokingHero) [inline R/cage_annotations.R]: space

File: R/feature_helpers.R [hunk 8]
```diff
@@ -106,6 +106,23 @@ fpkm_calc <- function(counts, lengthSize, librarySize){
...
+#' @param tx a GRangesList of transcripts containing grl
+#' @param faFile a FaFile from the fasta file, see ?FaFile.
+#'  Can also be path to fastaFile with fai file in same dir.
+#' @param groupBy (NULL) column to group grl by, if NULL group by names(gr)
+#' @return a character vector of start regions
+startRegionString <- function(grl, tx, faFile, groupBy = NULL){
+  gr <- startSites(grl, TRUE, TRUE, TRUE)
+
+  grl <- groupGRangesBy(windowPerGroup(grl, tx, 20, 20), groupBy)
+
```
Comment (JokingHero) [inline R/feature_helpers.R]: extra lines are for what really?

File: R/feature_helpers.R [hunk 9]
```diff
@@ -166,3 +183,88 @@ validExtension <- function(extension, cageFiveUTRs) {
...
+#' @param average a logical (FALSE), take average over coverage in all grl ?
+#' @param pShifted a logical (TRUE), are riboseq reads p-shifted?
+#' @param keep.names a logical(FALSE), only applies when onlyProportion
+#'  is TRUE.
+#' @return a data.frame with lengths by coverage / vector of proportions
+#' @importFrom BiocGenerics Reduce
+#'
+riboTISCoverageProportion <- function(grl, tx, footprints,
+                                      onlyProportion = FALSE, average = FALSE,
+                                        pShifted = TRUE, keep.names = FALSE){
```
Comment (JokingHero) [inline R/feature_helpers.R]: formatting, space between ) {, you could move outside the upStart and downStop params to get more flexible function

File: R/feature_methods2.R [hunk 10]
```diff
@@ -1,3 +1,72 @@
+#' Get initiation score for a GRangesList of ORFs
+#'
+#' initiationScore tries to check how much each TIS region resembles, the
+#' average of the CDS TIS regions.
+#'
```
Comment (JokingHero) [inline R/feature_methods2.R]: add citation to adams paper

File: R/shift_footprints.R [hunk 11]
```diff
@@ -140,35 +141,44 @@ detectRibosomeShifts <- function(
...
+
 
   txNames <- txNamesWithLeaders(txdb, minFiveUTR = minFiveUTR,
                                 minCDS = minCDS, minThreeUTR = minThreeUTR)
   if (length(txNames) == 0) stop("No transcript has leaders and trailers of",
                                  " specified minFiveUTR, minCDS, minThreeUTR")
   cds <- GenomicFeatures::cdsBy(txdb, by = "tx", use.names = TRUE)[txNames]
 
+  # reduce data set to only matching seqlevels
+  cds <- keepSeqlevels(cds, unique(seqnames(footprints))[unique(seqnames(
```
Comment (JokingHero) [inline R/shift_footprints.R]: format this properly

File: tests/testthat/test_features.R [hunk 12]
```diff
@@ -315,3 +314,29 @@ test_that("computeFeatures works as intended", {
   expect_equal(dt, featureExamples) # should be equal to saved version
 })
 
+test_that("initiationScore works as intended", {
+
```
Comment (JokingHero) [inline tests/testthat/test_features.R]: good

Comment (JokingHero) [reply tests/testthat/test_features.R]: add scaling to -1:1 for the function

File: R/feature_methods2.R (hunk 10 above)
Comment (Roleren) [reply R/feature_methods2.R]: Already linked

File: tests/testthat/test_features.R (hunk 12 above)
Comment (Roleren) [reply tests/testthat/test_features.R]: done

File: R/feature_helpers.R [hunk 13]
```diff
@@ -166,3 +181,88 @@ validExtension <- function(extension, cageFiveUTRs) {
...
+#' @param average a logical (FALSE), take average over coverage in all grl ?
+#' @param pShifted a logical (TRUE), are riboseq reads p-shifted?
+#' @param keep.names a logical(FALSE), only applies when onlyProportion
+#'  is TRUE.
+#' @return a data.frame with lengths by coverage / vector of proportions
+#' @importFrom BiocGenerics Reduce
+#'
+riboTISCoverageProportion <- function(grl, tx, footprints,
+                                      onlyProportion = FALSE, average = FALSE,
+                                        pShifted = TRUE, keep.names = FALSE) {
```
Comment (JokingHero) [inline R/feature_helpers.R]: fix

File: R/gr_helpers.R [hunk 14]
```diff
@@ -61,22 +61,33 @@ groupGRangesBy <- function(gr, other = NULL) {
...
 
   if (is(reads, "GRanges")) {
     rfpWidth <- width(reads)
     is.one_based <- all(as.integer(rfpWidth) == rep(1, length(rfpWidth)))
     if (is.one_based ) {
       if (is.null(reads$score)) {
-        message("All widths are 1, If ribo-seq is p-shifted, ",
-                "score column should contain widths of read, ",
-                "will continue using 1-widths")
+        if(is.null(reads$size)){
```
Comment (JokingHero) [inline R/gr_helpers.R]: space

File: R/ranges_helpers.R [hunk 15]
```diff
@@ -170,6 +170,26 @@ txSeqsFromFa <- function(grl, faFile, is.sorted = FALSE) {
...
+#'
+#' If downstreamFrom is 20, it means the window will start -20 downstream of
+#' gr start site. It will keep exon structure of tx
+#' @param gr a GRanges object (startSites and others, must be single point)
+#' @param tx a GRangesList of transcripts or (container region)
+#' @param downstream an integer, relative region to get downstream from
+#' @param upstream an integer vector, relative region to get upstream from.
+#' @return a GRanges/GRangesList object if exon/introns
+#'
+windowPerGroup <- function(gr, tx, downstream = 0L, upstream = 0L){
```
Comment (JokingHero) [inline R/ranges_helpers.R]: space

File: R/ranges_helpers.R [hunk 16]
```diff
@@ -455,6 +475,43 @@ extendLeaders <- function(grl, extension = 1000, cds = NULL) {
...
+#' @examples
+#' ORF <- GRanges(seqnames = "1",
+#'                ranges = IRanges(start = c(1, 10, 20),
+#'                                 end = c(5, 15, 25)),
+#'                strand = "+")
+#' names(ORF) <- "tx1"
+#' reads <- GRanges("1", IRanges(25, 25), "+")
+#' overlapsToCoverage(ORF, reads)
+#'
+overlapsToCoverage <- function(gr, reads, keep.names = TRUE, type = NULL){
```
Comment (JokingHero) [inline R/ranges_helpers.R]: space

File: R/shift_footprints.R (hunk 11 above)
Comment (JokingHero) [reply R/shift_footprints.R]: break it apart

**Pull request merged**